### ISSUE TITLE
BGM plays at full volume

### ISSUE THREAD

**Issue by Hareluya79** [wizzomafizzo/MiSTer_BGM#11]
Hi, thanks for this wonderful feature, it really makes the main pop!!
However music is very loud, 2 notch down from the Max Global volume makes it perfect but also causes the sound to be very low in cores, is there anyway to reduce the default volume to a comfortable level?
Much appreciate for your effort in this feature~

**Comment by KingKannibal** [comment]
You can permanently lower the gain on an mp3 file using foobar. Let me know if you need help doing this and I'll write up a quick guide for you.

**Comment by wizzomafizzo** [comment]
Yes I agree normalising the volume on all your tracks is probably the better solution. I'd be happy to add how to do that to the readme. I only know how to do it with ffmpeg

I'll investigate a global volume setting but I don't want to actually touch the volume of the mister, rather do it through the apps. I'm using 4 separate apps to play files at this point though so they'd all need to offer a volume feature. I'll see

**Comment by KingKannibal** [comment]
> Yes I agree normalising the volume on all your tracks is probably the better solution. I'd be happy to add how to do that to the readme. I only know how to do it with ffmpeg
> 
> I'll investigate a global volume setting but I don't want to actually touch the volume of the mister, rather do it through the apps. I'm using 4 separate apps to play files at this point though so they'd all need to offer a volume feature. I'll see

Would it be possible to add support for replay-gain? I noticed you included wav and vgm support. Having raw vgm files would cut down on space immensely but music in this format tends to be extremely loud and replay-gain helps so much with this.

**Comment by wizzomafizzo** [comment]
Can you explain what that involves? For the life of me I cannot find any documentation for vgmplay. I'm willing to add an exception for vgm files to adjust the volume

**Comment by KingKannibal** [comment]
That's also out of my league. All I know about it is that it applies a lower gain that makes music not sound overbearingly loud.
https://en.wikipedia.org/wiki/ReplayGain

**Comment by Hareluya79** [comment]
Hi All, I'm a Mac user, at the moment I'm using https://www.onlineconverter.com/increase-mp3-volume and set to Decrease 70% conversion to get the output to a non overwhelming level. Hope this workaround suffice till a better solution comes up, thanks to everyone helping out for this amazing feature.

Sharing a good place to get amazing game musics for your menus~
https://downloads.khinsider.com/game-soundtracks/album/classic-video-game-music

**Comment by wizzomafizzo** [comment]
I've done some more research for this now but it doesn't look too good so far.

- Out of mpg123, ogg123, aplay and vgmplay only mpg123 supports a (marked as deprecated) feature to adjust gain. So assuming that feature even works, it means we can only adjust volume on mp3 files
- It may be possible to write a separate compiled "player" app that supports all these formats and also can adjust gain, but that is complicated by the vgm format which hasn't got libraries for other languages. vgm being one of the formats in particular we want to fix
- Perhaps it's possible to adjust gain on the fly by writing to these files, but that's also very complex and I want to avoid pulling in extra python libraries. It's even possible this would start butting up against the limits of the ARM chip
- System volume can't be changed with any native linux applications like amixer
- There exists a Volume.dat file on the filesystem that can adjust volume (SAM uses this) but its contents is only read during a core load which isn't flexible enough for our purposes
- I checked through the mister menu code, that's the only thing in the whole mister that seems to be able to change volume, and it doesn't let anything else interface with that. It would be possible to write a utility to do what menu does to change the volume but this low level code is a bit over my head, it would take me a long time to figure it out

So for now unfortunately I think the workaround will be to offer a little documentation on adjusting the volume for the different music files. There's a couple ways forward but they're relatively complex given how simple the problem is and I've only got so much time to work on stuff

**Comment by KingKannibal** [comment]
I was planning on writing an article to help people with this anyway regardless of how big their personal music libraries would be for BGM. The quick of it is that we will have to rely on MP3s because they are the only filetype I see that can have replay-gain information permanently applied negating the use of extra software or code for volume adjustment. The reason you want replay-gain in the first place is so that all of your music plays at a similar volume, quiet songs will be louder, loud songs won't be overpowering everything (a problem with many VGM files). Another thing, do you know if it's possible for metadata to be displayed? at the moment, my hopes for that are low so I decided to strip all of my music of metadata to further save on space.

**Comment by wizzomafizzo** [comment]
It could be shown in the control GUI but I'm pretty sure there's no way to show it anywhere in the menu osd

**Comment by Hackmodford** [comment]
I agree. Having the ability to adjust the music volume would be beneficial

**Comment by Toryalai1** [comment]
Now that controlling system volume is possible, @wizzomafizzo could this please be looked into?

**Comment by wizzomafizzo** [comment]
The latest version of BGM should support this now. See this section of the docs: https://github.com/wizzomafizzo/MiSTer_BGM#volume-control

And please turn your mister off and back on after updating so BGM fully restarts.

There were a couple issues with volumes not applying consistently but I think that's fixed now.
